Error with kernel-4.8.0 and 4.8.1

Patch for kernel 4.8 seems to be incomplete:

```
ERROR: "validate_xmit_skb_list" [drivers/net/imq.ko] undefined!
ERROR: "dev_hard_start_xmit" [drivers/net/imq.ko] undefined!
make[1]: *** [scripts/Makefile.modpost:91: __modpost] Error 1
make: *** [Makefile:1180: modules] Error 2
```

Both functions are defined in netdevice.h and implemented in net/core/dev.c, but they are not visible. 
I had to add `EXPORT_SYMBOL(validate_xmit_skb_list);` and `EXPORT_SYMBOL(dev_hard_start_xmit);`in dev.c

Well, my c/c++ knowlegde are basics, kernel experience - a big zero.
I just compared versions from 4.6 to 4.8 and dig on internet to understand notions new to me, so please advise me. The kernel is compiled, both bzImage and all modules, but it is safe ?

--- a/kernel/v4.x/linux-4.8-imq.diff
+++ b/kernel/v4.x/linux-4.8-imq.diff
@@ -1,6 +1,6 @@
 diff -Naupr linux-4.8_orig/drivers/net/imq.c linux-4.8/drivers/net/imq.c
 --- linux-4.8_orig/drivers/net/imq.c	1970-01-01 07:00:00.000000000 +0700
-+++ linux-4.8/drivers/net/imq.c	2016-10-03 07:59:56.140345020 +0700
++++ linux-4.8/drivers/net/imq.c	2016-10-15 10:54:31.829676979 +0700
 @@ -0,0 +1,904 @@
 +/*
 + *             Pseudo-driver for the intermediate queue device.
@@ -553,7 +553,7 @@ diff -Naupr linux-4.8_orig/drivers/net/imq.c linux-4.8/drivers/net/imq.c
 +{
 +	struct sk_buff *skb_orig, *skb, *skb_shared, *skb_popd;
 +	struct Qdisc *q;
-+	struct sk_buff **to_free;
++	struct sk_buff *to_free = NULL;
 +	struct netdev_queue *txq;
 +	spinlock_t *root_lock;
 +	int users;
@@ -612,7 +612,7 @@ diff -Naupr linux-4.8_orig/drivers/net/imq.c linux-4.8/drivers/net/imq.c
 +
 +	/* backup skb->cb, as qdisc layer will overwrite it */
 +	skb_save_cb(skb_shared);
-+	qdisc_enqueue_root(skb_shared, q, to_free); /* might kfree_skb */
++	qdisc_enqueue_root(skb_shared, q, &to_free); /* might kfree_skb */
 +	if (likely(atomic_read(&skb_shared->users) == users + 1)) {
 +		bool validate;
 +
@@ -908,7 +908,7 @@ diff -Naupr linux-4.8_orig/drivers/net/imq.c linux-4.8/drivers/net/imq.c
 +MODULE_ALIAS_RTNL_LINK("imq");
 diff -Naupr linux-4.8_orig/drivers/net/Kconfig linux-4.8/drivers/net/Kconfig
 --- linux-4.8_orig/drivers/net/Kconfig	2016-10-03 06:24:33.000000000 +0700
-+++ linux-4.8/drivers/net/Kconfig	2016-10-03 07:59:56.140345020 +0700
++++ linux-4.8/drivers/net/Kconfig	2016-10-15 10:54:31.829676979 +0700
 @@ -258,6 +258,125 @@ config RIONET_RX_SIZE
  	depends on RIONET
  	default "128"
@@ -1037,7 +1037,7 @@ diff -Naupr linux-4.8_orig/drivers/net/Kconfig linux-4.8/drivers/net/Kconfig
  	depends on INET
 diff -Naupr linux-4.8_orig/drivers/net/Makefile linux-4.8/drivers/net/Makefile
 --- linux-4.8_orig/drivers/net/Makefile	2016-10-03 06:24:33.000000000 +0700
-+++ linux-4.8/drivers/net/Makefile	2016-10-03 07:59:56.200344939 +0700
++++ linux-4.8/drivers/net/Makefile	2016-10-15 10:54:31.913010159 +0700
 @@ -11,6 +11,7 @@ obj-$(CONFIG_DUMMY) += dummy.o
  obj-$(CONFIG_EQUALIZER) += eql.o
  obj-$(CONFIG_IFB) += ifb.o
@@ -1048,7 +1048,7 @@ diff -Naupr linux-4.8_orig/drivers/net/Makefile linux-4.8/drivers/net/Makefile
  obj-$(CONFIG_MII) += mii.o
 diff -Naupr linux-4.8_orig/include/linux/imq.h linux-4.8/include/linux/imq.h
 --- linux-4.8_orig/include/linux/imq.h	1970-01-01 07:00:00.000000000 +0700
-+++ linux-4.8/include/linux/imq.h	2016-10-03 07:59:56.200344939 +0700
++++ linux-4.8/include/linux/imq.h	2016-10-15 10:54:31.913010159 +0700
 @@ -0,0 +1,13 @@
 +#ifndef _IMQ_H
 +#define _IMQ_H
@@ -1065,7 +1065,7 @@ diff -Naupr linux-4.8_orig/include/linux/imq.h linux-4.8/include/linux/imq.h
 +
 diff -Naupr linux-4.8_orig/include/linux/netdevice.h linux-4.8/include/linux/netdevice.h
 --- linux-4.8_orig/include/linux/netdevice.h	2016-10-03 06:24:33.000000000 +0700
-+++ linux-4.8/include/linux/netdevice.h	2016-10-03 07:59:56.200344939 +0700
++++ linux-4.8/include/linux/netdevice.h	2016-10-15 10:54:31.913010159 +0700
 @@ -3602,6 +3602,19 @@ static inline void netif_tx_unlock_bh(st
  	}						\
  }
@@ -1088,7 +1088,7 @@ diff -Naupr linux-4.8_orig/include/linux/netdevice.h linux-4.8/include/linux/net
  	unsigned int i;
 diff -Naupr linux-4.8_orig/include/linux/netfilter/xt_IMQ.h linux-4.8/include/linux/netfilter/xt_IMQ.h
 --- linux-4.8_orig/include/linux/netfilter/xt_IMQ.h	1970-01-01 07:00:00.000000000 +0700
-+++ linux-4.8/include/linux/netfilter/xt_IMQ.h	2016-10-03 07:59:56.200344939 +0700
++++ linux-4.8/include/linux/netfilter/xt_IMQ.h	2016-10-15 10:54:31.913010159 +0700
 @@ -0,0 +1,9 @@
 +#ifndef _XT_IMQ_H
 +#define _XT_IMQ_H
@@ -1101,7 +1101,7 @@ diff -Naupr linux-4.8_orig/include/linux/netfilter/xt_IMQ.h linux-4.8/include/li
 +
 diff -Naupr linux-4.8_orig/include/linux/netfilter_ipv4/ipt_IMQ.h linux-4.8/include/linux/netfilter_ipv4/ipt_IMQ.h
 --- linux-4.8_orig/include/linux/netfilter_ipv4/ipt_IMQ.h	1970-01-01 07:00:00.000000000 +0700
-+++ linux-4.8/include/linux/netfilter_ipv4/ipt_IMQ.h	2016-10-03 07:59:56.200344939 +0700
++++ linux-4.8/include/linux/netfilter_ipv4/ipt_IMQ.h	2016-10-15 10:54:31.913010159 +0700
 @@ -0,0 +1,10 @@
 +#ifndef _IPT_IMQ_H
 +#define _IPT_IMQ_H
@@ -1115,7 +1115,7 @@ diff -Naupr linux-4.8_orig/include/linux/netfilter_ipv4/ipt_IMQ.h linux-4.8/incl
 +
 diff -Naupr linux-4.8_orig/include/linux/netfilter_ipv6/ip6t_IMQ.h linux-4.8/include/linux/netfilter_ipv6/ip6t_IMQ.h
 --- linux-4.8_orig/include/linux/netfilter_ipv6/ip6t_IMQ.h	1970-01-01 07:00:00.000000000 +0700
-+++ linux-4.8/include/linux/netfilter_ipv6/ip6t_IMQ.h	2016-10-03 07:59:56.200344939 +0700
++++ linux-4.8/include/linux/netfilter_ipv6/ip6t_IMQ.h	2016-10-15 10:54:31.913010159 +0700
 @@ -0,0 +1,10 @@
 +#ifndef _IP6T_IMQ_H
 +#define _IP6T_IMQ_H
@@ -1129,7 +1129,7 @@ diff -Naupr linux-4.8_orig/include/linux/netfilter_ipv6/ip6t_IMQ.h linux-4.8/inc
 +
 diff -Naupr linux-4.8_orig/include/linux/skbuff.h linux-4.8/include/linux/skbuff.h
 --- linux-4.8_orig/include/linux/skbuff.h	2016-10-03 06:24:33.000000000 +0700
-+++ linux-4.8/include/linux/skbuff.h	2016-10-03 07:59:56.200344939 +0700
++++ linux-4.8/include/linux/skbuff.h	2016-10-15 10:54:31.913010159 +0700
 @@ -39,6 +39,10 @@
  #include <linux/in6.h>
  #include <linux/if_packet.h>
@@ -1197,7 +1197,7 @@ diff -Naupr linux-4.8_orig/include/linux/skbuff.h linux-4.8/include/linux/skbuff
  	nf_bridge_get(src->nf_bridge);
 diff -Naupr linux-4.8_orig/include/net/netfilter/nf_queue.h linux-4.8/include/net/netfilter/nf_queue.h
 --- linux-4.8_orig/include/net/netfilter/nf_queue.h	2016-10-03 06:24:33.000000000 +0700
-+++ linux-4.8/include/net/netfilter/nf_queue.h	2016-10-03 07:59:56.200344939 +0700
++++ linux-4.8/include/net/netfilter/nf_queue.h	2016-10-15 10:54:31.913010159 +0700
 @@ -31,6 +31,12 @@ struct nf_queue_handler {
  void nf_register_queue_handler(struct net *net, const struct nf_queue_handler *qh);
  void nf_unregister_queue_handler(struct net *net);
@@ -1213,7 +1213,7 @@ diff -Naupr linux-4.8_orig/include/net/netfilter/nf_queue.h linux-4.8/include/ne
  void nf_queue_entry_release_refs(struct nf_queue_entry *entry);
 diff -Naupr linux-4.8_orig/include/net/pkt_sched.h linux-4.8/include/net/pkt_sched.h
 --- linux-4.8_orig/include/net/pkt_sched.h	2016-10-03 06:24:33.000000000 +0700
-+++ linux-4.8/include/net/pkt_sched.h	2016-10-03 07:59:56.200344939 +0700
++++ linux-4.8/include/net/pkt_sched.h	2016-10-15 10:54:31.913010159 +0700
 @@ -105,6 +105,8 @@ int sch_direct_xmit(struct sk_buff *skb,
  
  void __qdisc_run(struct Qdisc *q);
@@ -1225,7 +1225,7 @@ diff -Naupr linux-4.8_orig/include/net/pkt_sched.h linux-4.8/include/net/pkt_sch
  	if (qdisc_run_begin(q))
 diff -Naupr linux-4.8_orig/include/net/sch_generic.h linux-4.8/include/net/sch_generic.h
 --- linux-4.8_orig/include/net/sch_generic.h	2016-10-03 06:24:33.000000000 +0700
-+++ linux-4.8/include/net/sch_generic.h	2016-10-03 07:59:56.200344939 +0700
++++ linux-4.8/include/net/sch_generic.h	2016-10-15 10:54:31.913010159 +0700
 @@ -510,6 +510,13 @@ static inline int qdisc_enqueue(struct s
  	return sch->enqueue(skb, sch, to_free);
  }
@@ -1242,7 +1242,7 @@ diff -Naupr linux-4.8_orig/include/net/sch_generic.h linux-4.8/include/net/sch_g
  	return q->flags & TCQ_F_CPUSTATS;
 diff -Naupr linux-4.8_orig/include/uapi/linux/netfilter.h linux-4.8/include/uapi/linux/netfilter.h
 --- linux-4.8_orig/include/uapi/linux/netfilter.h	2016-10-03 06:24:33.000000000 +0700
-+++ linux-4.8/include/uapi/linux/netfilter.h	2016-10-03 07:59:56.203678268 +0700
++++ linux-4.8/include/uapi/linux/netfilter.h	2016-10-15 10:54:31.949676757 +0700
 @@ -14,7 +14,8 @@
  #define NF_QUEUE 3
  #define NF_REPEAT 4
@@ -1255,7 +1255,7 @@ diff -Naupr linux-4.8_orig/include/uapi/linux/netfilter.h linux-4.8/include/uapi
   * number or errno values. Not nice, but better than additional function
 diff -Naupr linux-4.8_orig/net/core/dev.c linux-4.8/net/core/dev.c
 --- linux-4.8_orig/net/core/dev.c	2016-10-03 06:24:33.000000000 +0700
-+++ linux-4.8/net/core/dev.c	2016-10-03 07:59:56.203678268 +0700
++++ linux-4.8/net/core/dev.c	2016-10-15 11:11:03.174515322 +0700
 @@ -141,6 +141,9 @@
  #include <linux/netfilter_ingress.h>
  #include <linux/sctp.h>
@@ -1266,9 +1266,40 @@ diff -Naupr linux-4.8_orig/net/core/dev.c linux-4.8/net/core/dev.c
  
  #include "net-sysfs.h"
  
+@@ -2911,7 +2914,12 @@ static int xmit_one(struct sk_buff *skb,
+ 	unsigned int len;
+ 	int rc;
+ 
++#if defined(CONFIG_IMQ) || defined(CONFIG_IMQ_MODULE)
++	if ((!list_empty(&ptype_all) || !list_empty(&dev->ptype_all)) &&
++		!(skb->imq_flags & IMQ_F_ENQUEUE))
++#else
+ 	if (!list_empty(&ptype_all) || !list_empty(&dev->ptype_all))
++#endif
+ 		dev_queue_xmit_nit(skb, dev);
+ 
+ 	len = skb->len;
+@@ -2950,6 +2958,8 @@ out:
+ 	return skb;
+ }
+ 
++EXPORT_SYMBOL(dev_hard_start_xmit);
++
+ static struct sk_buff *validate_xmit_vlan(struct sk_buff *skb,
+ 					  netdev_features_t features)
+ {
+@@ -3036,6 +3046,8 @@ struct sk_buff *validate_xmit_skb_list(s
+ 	return head;
+ }
+ 
++EXPORT_SYMBOL(validate_xmit_skb_list);
++
+ static void qdisc_pkt_len_init(struct sk_buff *skb)
+ {
+ 	const struct skb_shared_info *shinfo = skb_shinfo(skb);
 diff -Naupr linux-4.8_orig/net/core/skbuff.c linux-4.8/net/core/skbuff.c
 --- linux-4.8_orig/net/core/skbuff.c	2016-10-03 06:24:33.000000000 +0700
-+++ linux-4.8/net/core/skbuff.c	2016-10-03 07:59:56.203678268 +0700
++++ linux-4.8/net/core/skbuff.c	2016-10-15 10:54:32.023009956 +0700
 @@ -82,6 +82,87 @@ struct kmem_cache *skbuff_head_cache __r
  static struct kmem_cache *skbuff_fclone_cache __read_mostly;
  int sysctl_max_skb_frags __read_mostly = MAX_SKB_FRAGS;
@@ -1413,7 +1444,7 @@ diff -Naupr linux-4.8_orig/net/core/skbuff.c linux-4.8/net/core/skbuff.c
  /**
 diff -Naupr linux-4.8_orig/net/ipv6/ip6_output.c linux-4.8/net/ipv6/ip6_output.c
 --- linux-4.8_orig/net/ipv6/ip6_output.c	2016-10-03 06:24:33.000000000 +0700
-+++ linux-4.8/net/ipv6/ip6_output.c	2016-10-03 07:59:56.203678268 +0700
++++ linux-4.8/net/ipv6/ip6_output.c	2016-10-15 10:54:32.079676517 +0700
 @@ -65,9 +65,6 @@ static int ip6_finish_output2(struct net
  	struct in6_addr *nexthop;
  	int ret;
@@ -1440,7 +1471,7 @@ diff -Naupr linux-4.8_orig/net/ipv6/ip6_output.c linux-4.8/net/ipv6/ip6_output.c
  			    ip6_finish_output,
 diff -Naupr linux-4.8_orig/net/netfilter/core.c linux-4.8/net/netfilter/core.c
 --- linux-4.8_orig/net/netfilter/core.c	2016-10-03 06:24:33.000000000 +0700
-+++ linux-4.8/net/netfilter/core.c	2016-10-03 07:59:56.203678268 +0700
++++ linux-4.8/net/netfilter/core.c	2016-10-15 10:54:32.079676517 +0700
 @@ -311,10 +311,14 @@ next_hook:
  		ret = NF_DROP_GETERR(verdict);
  		if (ret == 0)
@@ -1460,7 +1491,7 @@ diff -Naupr linux-4.8_orig/net/netfilter/core.c linux-4.8/net/netfilter/core.c
  				goto next_hook;
 diff -Naupr linux-4.8_orig/net/netfilter/Kconfig linux-4.8/net/netfilter/Kconfig
 --- linux-4.8_orig/net/netfilter/Kconfig	2016-10-03 06:24:33.000000000 +0700
-+++ linux-4.8/net/netfilter/Kconfig	2016-10-03 07:59:56.203678268 +0700
++++ linux-4.8/net/netfilter/Kconfig	2016-10-15 10:54:32.093009827 +0700
 @@ -805,6 +805,18 @@ config NETFILTER_XT_TARGET_LOG
  
  	  To compile it as a module, choose M here.  If unsure, say N.
@@ -1482,7 +1513,7 @@ diff -Naupr linux-4.8_orig/net/netfilter/Kconfig linux-4.8/net/netfilter/Kconfig
  	depends on NETFILTER_ADVANCED
 diff -Naupr linux-4.8_orig/net/netfilter/Makefile linux-4.8/net/netfilter/Makefile
 --- linux-4.8_orig/net/netfilter/Makefile	2016-10-03 06:24:33.000000000 +0700
-+++ linux-4.8/net/netfilter/Makefile	2016-10-03 07:59:56.203678268 +0700
++++ linux-4.8/net/netfilter/Makefile	2016-10-15 10:54:32.116343117 +0700
 @@ -115,6 +115,7 @@ obj-$(CONFIG_NETFILTER_XT_TARGET_CT) +=
  obj-$(CONFIG_NETFILTER_XT_TARGET_DSCP) += xt_DSCP.o
  obj-$(CONFIG_NETFILTER_XT_TARGET_HL) += xt_HL.o
@@ -1493,7 +1524,7 @@ diff -Naupr linux-4.8_orig/net/netfilter/Makefile linux-4.8/net/netfilter/Makefi
  obj-$(CONFIG_NETFILTER_XT_TARGET_NETMAP) += xt_NETMAP.o
 diff -Naupr linux-4.8_orig/net/netfilter/nf_internals.h linux-4.8/net/netfilter/nf_internals.h
 --- linux-4.8_orig/net/netfilter/nf_internals.h	2016-10-03 06:24:33.000000000 +0700
-+++ linux-4.8/net/netfilter/nf_internals.h	2016-10-03 07:59:56.203678268 +0700
++++ linux-4.8/net/netfilter/nf_internals.h	2016-10-15 10:54:32.116343117 +0700
 @@ -18,7 +18,7 @@ unsigned int nf_iterate(struct list_head
  
  /* nf_queue.c */
@@ -1505,7 +1536,7 @@ diff -Naupr linux-4.8_orig/net/netfilter/nf_internals.h linux-4.8/net/netfilter/
  
 diff -Naupr linux-4.8_orig/net/netfilter/nf_queue.c linux-4.8/net/netfilter/nf_queue.c
 --- linux-4.8_orig/net/netfilter/nf_queue.c	2016-10-03 06:24:33.000000000 +0700
-+++ linux-4.8/net/netfilter/nf_queue.c	2016-10-03 07:59:56.203678268 +0700
++++ linux-4.8/net/netfilter/nf_queue.c	2016-10-15 10:54:32.116343117 +0700
 @@ -27,6 +27,23 @@
   * receives, no matter what.
   */
@@ -1573,7 +1604,7 @@ diff -Naupr linux-4.8_orig/net/netfilter/nf_queue.c linux-4.8/net/netfilter/nf_q
  			   (verdict & NF_VERDICT_FLAG_QUEUE_BYPASS))
 diff -Naupr linux-4.8_orig/net/netfilter/xt_IMQ.c linux-4.8/net/netfilter/xt_IMQ.c
 --- linux-4.8_orig/net/netfilter/xt_IMQ.c	1970-01-01 07:00:00.000000000 +0700
-+++ linux-4.8/net/netfilter/xt_IMQ.c	2016-10-03 07:59:56.203678268 +0700
++++ linux-4.8/net/netfilter/xt_IMQ.c	2016-10-15 10:54:32.116343117 +0700
 @@ -0,0 +1,72 @@
 +/*
 + * This target marks packets to be enqueued to an imq device
@@ -1649,7 +1680,7 @@ diff -Naupr linux-4.8_orig/net/netfilter/xt_IMQ.c linux-4.8/net/netfilter/xt_IMQ
 +
 diff -Naupr linux-4.8_orig/net/sched/sch_generic.c linux-4.8/net/sched/sch_generic.c
 --- linux-4.8_orig/net/sched/sch_generic.c	2016-10-03 06:24:33.000000000 +0700
-+++ linux-4.8/net/sched/sch_generic.c	2016-10-03 07:59:56.207011597 +0700
++++ linux-4.8/net/sched/sch_generic.c	2016-10-15 10:54:32.139676407 +0700
 @@ -154,6 +154,14 @@ bulk:
  	return skb;
  }

